### ISSUE TITLE
[Client] Negative example for object detection

### ISSUE THREAD

**Issue by michael-may** [cloud-annotations/training#54]
Is there an easy way to provide negative samples for object detection models?

It looks like you can kind of work around it by adding boxes in the cloud annotations tool and then deleting them. They end up in the _annotations.json file without any labels applied. It works, but it's not super intuitive. I realize this might be out of the scope of this repository, so let me know if this question is best directed elsewhere.

**Comment by bourdakos1** [comment]
This is the perfect place to ask these questions :)

This feature is something I haven’t really thought of before. The solution you are using was a bug and it should be fixed now, and if not, please let me know. 

Actual solution wise, do you have any ideas? The simplest way could just be a “mark as negative” button that adds the image to the annotations without any labels and it could show up as “negatives” on the sidebar. My only issue with this is that I don’t like adding a bunch of buttons because the tool would get rather cluttered and easily become confusing to non-power users. A slightly more hidden approach could have a reserve work labels, like “negative” and if you choose that label your only option is to make the image as a negative, and not draw boxes.

Let me know what you think of those 2 options or if you have any better ideas

For now if you need a work around you can run:
```
cacli train path/to/file.zip
```
With a custom version of the default training script which can be could under releases.
Just change it from only downloading labeled images, to downloading all images

**Comment by michael-may** [comment]
I think adding negative images is probably more of a bulk operation most of the time, right?

I noticed the classification version of the tool lets you select multiple images at once, but the annotation version doesn't right now. Maybe an option could appear in a toolbar at the top (similar to the other version of the tool) only if the user has multiple images selected? that would keep the right-hand bar nice and clean.

![image](https://user-images.githubusercontent.com/5215114/56427470-96b10e80-6281-11e9-93a1-691d58405d0c.png)

Also, I can open another issue if it's a bug, but is there a supported method of having multiple users work on annotating at the same time? Even with different API keys, users seem to step over each other's changes. I'm assuming it's probably just because the changes are all written to the same _annotations.json?

**Comment by bourdakos1** [comment]
Yea, I think a multi select system would be the best solution, that way you could also bulk delete images.

And yea, let’s open another issue where we can discuss multi-user editing

**Comment by oumingyang** [comment]
Hi, when i train the object-detection model with the training_video. need I prepare some unlabeled object ?

**Comment by sachinshelke** [comment]
**Is current implementation considered unlabelled images as negative?**

I did trained model with too many steps about 200K for 7 classes and around 1K images for each classes.

But still i'm getting some unacceptable results as non - relevant images are detected with 99% accuracy.
Could you please suggest any way to avoid this.
Some configuration id changed for TFWrapper for 
```
iou_threshold = 0.95
score_threshold = 0.95
```
But still no luck.

**Comment by bourdakos1** [comment]
At the moment any unlabeled image is **NOT** included in the trading dataset. Hopefully soon in the tool you'll be mark unlabeled images to be included in the training set. For now you can download the training.zip and modify it to download all images from the bucket when training

**Comment by sachinshelke** [comment]
Thanks Nick for updates.
But i **guess** we have to add all unlabelled imaged in tfrecord as well. Is it so?

**Comment by bourdakos1** [comment]
Correct, but if I remember correctly I think as long as the images are downloaded they are added to the tfrecord, but it could be that it needs to be listed in the annotations.json as well.

**Comment by michael-may** [comment]
We train our data locally right now, so it might be different if you use the cacli, but here's what has worked for us so far:

The workaround we came up with (for now) was to create a class called `_negative_` and draw random boxes to make sure they're pulled down with the training data.

![Screen Shot 2019-10-15 at 11 15 38 AM](https://user-images.githubusercontent.com/5215114/66849786-1f60ec80-ef3d-11e9-9272-8e5dd4dd04fa.png)

Then we modified [this file](https://github.com/cloud-annotations/training/blob/master/trainer/src/data/prepare_data_object_detection.py) as below:

![Annotation 2019-10-15 111143](https://user-images.githubusercontent.com/5215114/66850069-aca44100-ef3d-11e9-9056-651bde06622a.png)
(The line numbers on the right are off due to comments higher in the file)

This lets the images pass through to the training data without any labels, which the object detection models will interpret as a negative sample.

**Comment by bourdakos1** [comment]
If other issues don’t pop up, I should have a solution to this by the end of the week 🤞

**Comment by bourdakos1** [comment]
This is what I have UI wise so far, thoughts @michael-may @sachinshelke?
<img width="1552" alt="Screen Shot 2019-10-15 at 6 41 59 PM" src="https://user-images.githubusercontent.com/4212769/66875260-acc33180-ef7b-11e9-84ff-f28ab1d3c03b.png">

Not sure if it makes sense to have a dedicated "Mark as negative" option? (marking any image with a label would include it in the training set, but all would be treated as "Negative" unless an actual box is marked)

**Comment by bourdakos1** [comment]
You can try it out at: https://stagingannotations.us-east.containers.appdomain.cloud
(this is the development site, so things might be buggy or broken from time to time)

**Comment by bourdakos1** [comment]
Does this make sense for your use cases?
```python
for annotation in annotations[example]:
    if 'x' in annotation and 'x2' in annotation and 'y' in annotation and 'y2' in annotation:
        xmins.append(annotation['x'])
        xmaxs.append(annotation['x2'])
        ymins.append(annotation['y'])
        ymaxs.append(annotation['y2'])
        classes_text.append(annotation['label'].encode('utf8'))
        classes.append(1) # temporary, I need to assign labels to actual ids
        difficult_obj.append(0)
        truncated.append(0)
        poses.append(''.encode('utf8'))
```

**Comment by bourdakos1** [comment]
Yea, I’m keeping both “mark as negative” and “mark as >”

attaching any label to the image without drawing boxes would, technically, treat it as negative, because it would be included in the dataset because it’s labeled, but doesn’t actually have any labels as far as object detection is concerned 

The feature is live now so I’m gonna go ahead and close this, but feel free to reopen if it doesn’t meet your needs :)

**Comment by sampathkumaran90** [comment]
Hi, I am facing similar issue. When i set min_negatives_per_image to 0 and not set any negative images, i get prediction correct for positive images but getting a lot of false positives. When i add a bunch of negative images say 50% positive and 50% negative and set min_negatives_per_image to 10, the results are not good. I am getting more false negatives as well as false positives. Anyone could help me on this issue... I am using ssd mobilenet v2 or finetuning

**Comment by mahamatnoumai** [comment]
> Hi, I am facing similar issue. When i set min_negatives_per_image to 0 and not set any negative images, i get prediction correct for positive images but getting a lot of false positives. When i add a bunch of negative images say 50% positive and 50% negative and set min_negatives_per_image to 10, the results are not good. I am getting more false negatives as well as false positives. Anyone could help me on this issue... I am using ssd mobilenet v2 or finetuning

Have you managed to fix this issues, could you please share your experiences if you fixed this issues in any way possible. Tq.
